### PR TITLE
documentation release v1

### PR DESCRIPTION
**Pull request by Titus-von-Koeller**

### CODE REVIEW

File: docs/source/optimizers.mdx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+#### Benefits:
+
+- Designed to support more aggressive quantization strategies without compromising training stability.
+- Helps in achieving stable training outcomes, particularly important for models dealing with diverse and complex language data.
+
+## Usage
+
+Some more examples of how you can replace your old optimizer with the 8-bit optimizer:
+
+```diff
```
Comment (younesbelkada) [inline docs/source/optimizers.mdx:79]: Here I would use `diff` to better highlight to users what are the needed modifs

File: docs/source/optimizers.mdx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+```
+
+The arguments passed are the same as standard Adam. For NLP models we recommend also to use the StableEmbedding layers which improves results and helps with stable 8-bit optimization.
+
+## Overview of supported 8-bit optimizers
+
+TOOD: List here all optimizers in `bitsandbytes/optim/__init__.py`
+TODO (future) document all optimizers as we did for Linear4bit / Linear8bitLt classes
+
+## Overview of expected gradients
```
Comment (younesbelkada) [inline docs/source/optimizers.mdx:34]: What this means exactly? Can you elaborate more about "expected gradients"? I also thing this is too in-depth and should be moved below

File: docs/source/optimizers.mdx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+<div style="text-align: center">
+<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bitsandbytes/optimizer_comparison.png", width="50%">
+</div>
+
+<div style="text-align: center">
+<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bitsandbytes/optimizer_largest_model.png", width="50%">
+</div>
```
Comment (younesbelkada) [inline docs/source/optimizers.mdx:42]: @Titus-von-Koeller this is how you add images (it uses HTML code)

File: docs/source/moduletree.mdx [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Module tree overview
```
Comment (younesbelkada) [inline docs/source/moduletree.mdx:1]: Is this file really needed?

File: docs/source/introduction.mdx
Comment (younesbelkada) [inline docs/source/introduction.mdx:1]: This file should stay as `index.mdx` and be referred as the first page / index page that briefly presents what bitsandbytes is, in one paragraph & lists the available documentation sections (quickstart, quantization, optimizers, integrations, resources)

File: docs/source/introduction.mdx [hunk 5]
```diff
@@ -1,52 +1,14 @@
...
+# `bitsandbytes`
 
+The `bitsandbytes` library is a lightweight Python wrapper around CUDA custom functions, in particular 8-bit optimizers, matrix multiplication (LLM.int8()), and quantization functions.
 
+There are ongoing efforts to support further hardware backends, i.e. Intel CPU + GPU, AMD GPU, Apple Silicon. Windows support is on its way as well.
+The library includes quantization primitives for 8-bit & 4-bit operations, through `bitsandbytes.nn.Linear8bitLt` and `bitsandbytes.nn.Linear4bit` and 8bit optimizers through `bitsandbytes.optim` module.
 
-Resources:
-- [8-bit Optimizer Paper](https://arxiv.org/abs/2110.02861) --  [Video](https://www.youtube.com/watch?v=IxrlHAJtqKE) -- [Docs](https://bitsandbytes.readthedocs.io/en/latest/)
+**Using 8-bit optimizers**:
```
Comment (younesbelkada) [inline docs/source/introduction.mdx:10]: I think this and below should be completely removed

File: docs/source/integrations.mdx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Transformers
+
+... TODO: to be filled out ...
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:3]: Here we point out to relevant doc sections in transformers / peft / Trainer + explain how these are integrated 

e.g. for transformers state that you can load any model in 8-bit / 4-bit precision, for PEFT, you can use QLoRA out of the box with `LoraConfig` + 4-bit base model, for Trainer: all bnb optimizers are supported by passing the correct string in `TrainingArguments` : https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/training_args.py#L134 

Few references: 

- transformers: https://huggingface.co/docs/transformers/quantization#bitsandbytes
- PEFT: https://huggingface.co/docs/peft/developer_guides/quantization

File: docs/source/faqs.mdx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# FAQs
```
Comment (younesbelkada) [inline docs/source/faqs.mdx:1]: Very good

File: docs/source/contributing.mdx [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Contributors guidelines
```
Comment (younesbelkada) [inline docs/source/contributing.mdx:1]: Very good !

File: docs/source/_toctree.yml [hunk 9]
```diff
@@ -1,8 +1,30 @@
...
+  - local: resources
+    title: Papers, related resources & how to cite
+  - local: faqs
+    title: FAQs (Frequently Asked Questions)
+- title: Contributors Guidelines
+  sections:
+  - local: contributing
+    title: Contributing
+  # - local: code_of_conduct
+  #   title: Code of Conduct
```
Comment (younesbelkada) [inline docs/source/_toctree.yml:30]:
```suggestion
```

File: docs/source/_toctree.yml [hunk 10]
```diff
@@ -1,8 +1,30 @@
-- sections:
-  - local: index
-    title: Bits & Bytes
+- title: Get started
+  sections:
+  - local: introduction
+    title: Introduction
```
Comment (younesbelkada) [inline docs/source/_toctree.yml:4]:
```suggestion
  - local: index
    title: index
```
No need for introduction, this should be converted to "index", see my comment below

File: bitsandbytes/nn/modules.py [hunk 11]
```diff
@@ -397,9 +441,50 @@ def maybe_rearrange_weight(state_dict, prefix, local_metadata, strict, missing_k
...
+        """
+        Initialize Linear8bitLt class.
+
+        Args:
+            input_features (`str`):
+                Number of input features of the linear layer.
+            output_features (`str`):
+                Number of output features of the linear layer.
+            bias (`bool`, defaults to `True`):
+                Whether the linear class uses the bias term as well.
```
Comment (younesbelkada) [inline bitsandbytes/nn/modules.py:485]: This is how you write docstring that is compatible with HF-doc builder auto-doc

**Review by younesbelkada — COMMENTED** [review]
Thanks for the huge work @Titus-von-Koeller !
I left few comments, we should be almost there to merge this !
Please also see the tips regarding uploading images, writing proper docstring so that things will be compatible with HF-doc builder's auto-doc. Let's also not blow up the documentation with the introduction, a simple index page that redirects users to the correct page should be sufficient. See for example: https://github.com/huggingface/trl/blob/main/docs/source/index.mdx on how to use relative links between different doc pages.

File: docs/source/introduction.mdx (hunk 5 above)
Comment (Titus-von-Koeller) [reply docs/source/introduction.mdx:10]: ok, deleted everything from here on down

Comment (Titus-von-Koeller) [reply docs/source/introduction.mdx:10]: even though this part here below is still relevant, maybe? I mean the last two bullets. Wdyt?

## Features

- 8-bit Matrix multiplication with mixed precision decomposition
- LLM.int8() inference
- 8-bit Optimizers: Adam, AdamW, RMSProp, LARS, LAMB, Lion (saves 75% memory)
- Stable Embedding Layer: Improved stability through better initialization, and normalization
- 8-bit quantization: Quantile, Linear, and Dynamic quantization
- Fast quantile estimation: Up to 100x faster than other algorithms

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: v0.1.15
+    rev: v0.2.0
```
Comment (younesbelkada) [inline .pre-commit-config.yaml:3]: Just want to make sure if this is intended

File: README.md [hunk 13]
```diff
@@ -4,10 +4,7 @@ The bitsandbytes is a lightweight wrapper around CUDA custom functions, in parti
```
Comment (younesbelkada) [inline README.md:6]:
```suggestion
Read more about bitsandbytes on its documentation section: https://huggingface.co/docs/bitsandbytes/main
```

File: bitsandbytes/nn/modules.py [hunk 14]
```diff
@@ -19,6 +19,42 @@
 
 
 class StableEmbedding(torch.nn.Embedding):
+    """
```
Comment (younesbelkada) [inline bitsandbytes/nn/modules.py:22]: Very impressive! Thanks @Titus-von-Koeller !

File: docs/source/algorithms.mdx [hunk 15]
```diff
@@ -0,0 +1,12 @@
+# Other algorithms
+_WIP: Still incomplete... Community contributions would be greatly welcome!_
+
+This is an overview of the algorithms in `bitsandbytes` that we think would also be useful as standalone entities.
```
Comment (younesbelkada) [inline docs/source/algorithms.mdx:4]:
```suggestion
This is an overview of the functional API in `bitsandbytes` that we think would also be useful as standalone entities.
```
I would rather call that a section about functional APIs no?

File: docs/source/compiling.mdx [hunk 16]
```diff
@@ -0,0 +1,41 @@
+# Compiling from Source[[compiling]]
```
Comment (younesbelkada) [inline docs/source/compiling.mdx:1]: Nice trick !

File: docs/source/installation.mdx [hunk 17]
```diff
@@ -5,6 +5,12 @@ Note currently `bitsandbytes` is only supported on CUDA GPU hardwares, support f
 <hfoptions id="OS system">
 <hfoption id="Linux">
 
+## Hardware requirements:
+ - LLM.int8(): NVIDIA Turing (RTX 20xx; T4) or Ampere GPU (RTX 30xx; A4-A100); (a GPU from 2018 or newer).
```
Comment (younesbelkada) [inline docs/source/installation.mdx:9]:
```suggestion
 - [LLM.int8()](https://arxiv.org/abs/2208.07339) & [QLoRA](https://arxiv.org/abs/2305.14314): NVIDIA Turing (RTX 20xx; T4) or Ampere GPU (RTX 30xx; A4-A100); (a GPU from 2018 or newer).
```

Comment (younesbelkada) [reply docs/source/installation.mdx:9]: I mentioned QLoRA as well

File: docs/source/installation.mdx [hunk 18]
```diff
@@ -5,6 +5,12 @@ Note currently `bitsandbytes` is only supported on CUDA GPU hardwares, support f
...
+ - 8-bit optimizers and quantization: NVIDIA Kepler GPU or newer (>=GTX 78X).
+
+Supported CUDA versions: 10.2 - 12.0  #TODO: check currently supported versions
```
Comment (younesbelkada) [inline docs/source/installation.mdx:12]:
```suggestion
Supported CUDA versions: 10.2 - 12.0 
```

File: docs/source/integrations.mdx [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+# Trainer for the optimizers
+
+... TODO: to be filled out ...
+
+Here we point out to relevant doc sections in transformers / peft / Trainer + very briefly explain how these are integrated:
+e.g. for transformers state that you can load any model in 8-bit / 4-bit precision, for PEFT, you can use QLoRA out of the box with `LoraConfig` + 4-bit base model, for Trainer: all bnb optimizers are supported by passing the correct string in `TrainingArguments` : https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/training_args.py#L134
+
+Few references:
+
+- transformers: https://huggingface.co/docs/transformers/quantization#bitsandbytes
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:18]:
```suggestion
- [transformers documentation]( https://huggingface.co/docs/transformers/quantization#bitsandbytes)
```

File: docs/source/integrations.mdx [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+- PEFT: https://huggingface.co/docs/peft/developer_guides/quantization
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:19]:
```suggestion
- [PEFT documentation](https://huggingface.co/docs/peft/developer_guides/quantization)
```

File: docs/source/integrations.mdx [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+# PEFT
+
+... TODO: to be filled out ...
+
+# Trainer for the optimizers
+
+... TODO: to be filled out ...
+
+Here we point out to relevant doc sections in transformers / peft / Trainer + very briefly explain how these are integrated:
+e.g. for transformers state that you can load any model in 8-bit / 4-bit precision, for PEFT, you can use QLoRA out of the box with `LoraConfig` + 4-bit base model, for Trainer: all bnb optimizers are supported by passing the correct string in `TrainingArguments` : https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/training_args.py#L134
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:14]:
```suggestion
e.g. for transformers state that you can load any model in 8-bit / 4-bit precision, for PEFT, you can use QLoRA out of the box with `LoraConfig` + 4-bit base model, for Trainer: all bnb optimizers are supported by passing the correct string in `TrainingArguments`'s `optim` attribute - e.g. (`paged_adamw_32bit`):
```

File: docs/source/integrations.mdx [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+
+Few references:
+
+- transformers: https://huggingface.co/docs/transformers/quantization#bitsandbytes
+- PEFT: https://huggingface.co/docs/peft/developer_guides/quantization
+
+# Blog posts
+
+- [Making LLMs even more accessible with bitsandbytes, 4-bit quantization and QLoRA](https://huggingface.co/blog/4bit-transformers-bitsandbytes)
+
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:24]:
```suggestion
- [A Gentle Introduction to 8-bit Matrix Multiplication for transformers at scale using Hugging Face Transformers, Accelerate and bitsandbytes](https://huggingface.co/blog/hf-bitsandbytes-integration)
```

File: docs/source/integrations.mdx [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+###
+
+For instructions how to use LLM.int8() inference layers in your own code, see the TL;DR above or for extended instruction see [this blog post](https://huggingface.co/blog/hf-bitsandbytes-integration).
```
Comment (younesbelkada) [inline docs/source/integrations.mdx:27]:
```suggestion
```
I would just mention the blogpost in the list of blogposts above

**Review by younesbelkada — APPROVED** [review]
Fantastic work @Titus-von-Koeller ! 🚀  I left few final comments that you can directly apply through GH suggestions ! Let's 🚢  it once we confirm the Doc build is green after applying the suggestions !
Thanks again for putting all your hardwork on this !
